pkg/monitor/agent.go

```diff
@@ -0,0 +1,131 @@
+	crdName := monitor.querier.GetNodeName()
+	agentCRD, err := monitor.client.ClusterinformationV1beta1().AntreaAgentInfos().Get(crdName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(2).Infof("Agent monitoring CRD named %s doesn't exist, will create one", crdName)
+		return nil
+	}
+	return agentCRD
+}
+
+func (monitor *agentMonitor) createAgentCRD() (*v1beta1.AntreaAgentInfo, error) {
+	agentCRD := monitor.querier.GetAgentInfo()
+	klog.V(2).Infof("Creating agent monitoring CRD %+v", agentCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaAgentInfos().Create(agentCRD)
+}
+
+// updateAgentCRD updates all the fields of existing monitoring CRD.
+func (monitor *agentMonitor) updateAgentCRD(agentCRD *v1beta1.AntreaAgentInfo) (*v1beta1.AntreaAgentInfo, error) {
+	agentCRD.Version = querier.GetVersion()
+	agentCRD.PodRef = querier.GetSelfPod()
+	agentCRD.NodeRef = querier.GetSelfNode(true, monitor.querier.GetNodeName())
+	agentCRD.NodeSubnet = []string{monitor.querier.GetNodeSubnet()}
+	ovsVersion := monitor.querier.GetOVSVersion()
+	// OVS version query will fail and return empty string when OVSDB connection is down.
+	// Only change OVS version when the query gets a valid version.
+	ovsConnected := ovsVersion != ""
+	if ovsConnected {
+		agentCRD.OVSInfo.Version = ovsVersion
+	}
+	agentCRD.OVSInfo.BridgeName = monitor.querier.GetOVSBridge()
+	agentCRD.OVSInfo.FlowTable = monitor.querier.GetOVSFlowTable()
+	agentCRD.NetworkPolicyControllerInfo = monitor.querier.GetNetworkPolicyControllerInfo()
+	agentCRD.LocalPodNum = int32(monitor.querier.GetInterfaceStore().GetContainerInterfaceNum())
+	agentCRD.AgentConditions = monitor.querier.GetAgentConditions(ovsConnected)
+	klog.V(2).Infof("Updating agent monitoring CRD %+v", agentCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaAgentInfos().Update(agentCRD)
+}
+
+// partialUpdateAgentCRD only updates the variables.
+func (monitor *agentMonitor) partialUpdateAgentCRD(agentCRD *v1beta1.AntreaAgentInfo) (*v1beta1.AntreaAgentInfo, error) {
+	// LocalPodNum, FlowTable, NetworkPolicyControllerInfo, OVSVersion and AgentConditions can be changed, so reset these fields.
+	agentCRD.LocalPodNum = int32(monitor.querier.GetInterfaceStore().GetContainerInterfaceNum())
+	agentCRD.OVSInfo.FlowTable = monitor.querier.GetOVSFlowTable()
+	agentCRD.NetworkPolicyControllerInfo = monitor.querier.GetNetworkPolicyControllerInfo()
+	ovsVersion := monitor.querier.GetOVSVersion()
+	// OVS version query will fail and return empty string when OVSDB connection is down.
+	// Only change OVS version when the query gets a valid version.
+	ovsConnected := ovsVersion != ""
+	if ovsConnected {
+		agentCRD.OVSInfo.Version = ovsVersion
+	}
+	agentCRD.AgentConditions = monitor.querier.GetAgentConditions(ovsConnected)
+	klog.V(2).Infof("Partially updating agent monitoring CRD %+v", agentCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaAgentInfos().Update(agentCRD)
+}
```

Comment:
The above 3 functions have many repeated code, and some get all information via one method `GetAgentInfo ` while some via a number of methods.
I think querier either provide a single method to return AgentInfo or provide a number of methods but know nothing about AgentInfo. No need to have two kind of methods for duplicate information.
If in former way, querier can just have a method
`GetAgentInfo(agentInfo *v1beta1.AntreaAgentInfo, partial bool)` which can be used by the above 3 functions.
what do you think @jianjuns ?

Reply:
I agree with you to let GetAgentInfo() have a param "partial". Then, for other methods, I think we can make them internal and used by GetAgentInfo().

Reply:
Code updated. Use GetAgentInfo() and GetControllerInfo() to get all info for create, update and partiallyUpdate. Turned other external methods into internal ones.
I keep these changes in a new commit for review. I will rebase it when code is ok to merge.